feat: ABB robot guide (#334)
aica-technology/api: docs/docs/examples/guides/abb-hardware-interface.md

                  \J6:=egm_condition
                  \MaxSpeedDeviation:=120.0;

      WHILE TRUE DO
          ! Start the EGM communication session.
          EGMRunJoint egm_id, EGM_STOP_HOLD, \J1 \J2 \J3 \J4 \J5 \J6 \CondTime:=5 \RampOutTime:=5 \PosCorrGain:=0;
          EGMStop egm_id, EGM_STOP_HOLD \RampOutTime:=5;
      ENDWHILE
      ! Release the EGM id.
      EGMReset egm_id;

  ERROR
      IF ERRNO = ERR_UDPUC_COMM THEN
          TPWrite "Communication timed out";
          TRYNEXT;
      ENDIF
  ENDPROC
ENDMODULE
```

</details>

## Hardware interface

:::tip

For best results, always set the rate of the hardware interface to 250 Hertz, which corresponds to the ABB-suggested
stable UDP data exchange limit.

:::

Returning to AICA Studio and the hardware interface, it is now possible to define the parameters and connect to the
robot. The majority of the hardware interface parameters enable connection to EGM and RWS:

<div class="text--center">
  <img src={abbHIParameters} alt="ABB Hardware interface parameters." />
</div>

- EGM port: the port that EGM uses to send commands (6511, the Remote port of the UDPUC device defined in RobotStudio).

> - RWS IP & port: address and port of the RWS (the address of the RobotStudio device or real robot, 192.168.137.1 and

Reviewer: I mentioned this above, I don't think users always know how to access the logs below easily, so might be better not to show them and instead work with the default values for ports? What do you think?

<div class="text--center">
  <img src={abbNewProject} alt="Create a new RobotStudio project." />
</div>

5. Select **Create** to create the new project.
6. In the window that pops up, in the **Options** tab, look for **EGM** and **RobotStudio Connect** and add them in the
   controller. Then select **Apply and Reset** to finalize.

:::note

The **3119-1 RobotStudio Connect** add-in is required to connect a controller to RobotStudio over a public network. For more
information, see the RobotStudio instruction manual. 

:::

<div class="text--center">
  <img src={abbAdditionalOptions} alt="Additional options in the RobotStudio project." />
</div>

:::tip

Another way of ensuring applications are set up correctly is to use the provided mock interfaces. The mock interface 
does not include a physics engine, or dynamics calculations, and merely sets the commands as state.

::: 

### Real robot controller

RobotStudio can be used to configure the address of the external control device. Navigate to the Controller tab and
select **Add Controller > Connect to Controller**. This will allow to detect and connect to the running controller in

> Omnicore, provided of course that the devices are on the same network.

Reviewer 2: ```suggestion
OmniCore, provided of course that the devices are on the same network.
```

and then restart.

:::

After connecting to the robot, the controller should be configured to accept commands from an external device.

1. Navigate to the Controller tab > Configuration > Communication > UDP Unicast Device, and add a new UDPUC device (or
   modify the existing one), configured as shown below. This is the device that will be running the AICA application,
   the external control device, so the address should be set accordingly.

<div class="text--center">
  <img src={abbControllerConfiguration} alt="Controller configuration settings." />
</div>

<div class="text--center">
  <img src={abbNewUDPUCDevice} alt="Add a new UDPUS device." />
</div>

2. The next step is enabling RWS connection. For the RobotStudio simulation, this requires either using a proxy or
   whitelisting the IP address of the device trying to access RWS, which is the device running AICA Core.
   The first approach is preferable and described analytically in a
   [RobotStudio forum post](https://forums.robotstudio.com/discussion/12082/using-robotwebservices-to-access-a-remote-virtual-controller)
   (read until the end and the last comment for a critical fix).
   :::note

>    Omnicore controllers and RobotWare 7.x versions by default listen on HTTPS and port 80 for RobotStudio and 443 for

Reviewer 2: ```suggestion
   OmniCore controllers and RobotWare 7.x versions by default listen on HTTPS and port 80 for RobotStudio and 443 for
```

import abbAdditionalOptions from './assets/abb-additional-options.png'
import abbNewProject from './assets/abb-new-project.png'
import abbInstallAddins from './assets/abb-install-addins.png'
import abbConnectController from './assets/abb-connect-controller.png'
import abbFirewallManager from './assets/abb-firewall-manager.png'

# ABB manipulators

ABB offers a wide range of industrial articulated manipulators, from compact 6-axis robots for small-part handling and
payloads of a few kilograms to heavy-duty models capable of lifting up to 800 kg. This guide aims to support the
integration of ABB robots in AICA System applications, from the mock interface and a simulated environment to the real
robot.

To use the ABB collection, add the latest version of `collections/abb` to your configuration in AICA Launcher, currently
supporting the following robot models:

- IRB 1010
- GoFa CRB 15000, 12 Kg

## General

In this section, some of the prerequisites to use the AICA ABB collection are introduced and discussed.

### EGM

ABB permits remote control of its manipulator range using the optional **Externally Guided Motion (EGM)** feature. EGM

> provides external devices with the ability to send commands and control ABB robotic arms, using Google's Protocol

Reviewer: Now that you added the warnings and notes, I don't mind if you want to switch the title of this section back to `General`.

Buffers (Protobuf) serialization library to transport information through UDP sockets. For more information, check out
the
[official product documentation](https://library.e.abb.com/public/344f15f0f43341eb944fe35279d9fa2e/3HAC073319+AM+Externally+Guided+Motion+RW6-en.pdf?x-sign=WlxgV7Vao27KV3d3hlsfaoykgctYqoA0F98ch89S%2FPEaGwQg47ou%2FioylQtzvLaV).

:::tip

For best results, always set the rate of the hardware interface to 250 Hertz, which corresponds to the ABB-suggested
stable UDP data exchange limit.

:::

> ### RWS
> 
> The second ABB feature that AICA system utilizes to connect to the robot is RobotWebServices (RWS). RWS is a platform

Reviewer: ```suggestion

### Robot Web Services

The second ABB feature that AICA system utilizes to connect to the robot is Robot Web Services (RWS). RWS is a platform
```

that enables developers to create applications that interact with the robot controller, using RESTful APIs that leverage
the HTTPS protocol. The hardware interface uses RWS for auxiliary functionality, such as starting/stopping the program
and the motors, and setting IOs. Setting up RWS in the simulator and the actual robot requires slightly different steps,
which will be explained in the following sections. More information can be found in the
[product reference page](https://developercenter.robotstudio.com/api/RWS).

### RAPID

RAPID is the programming language of ABB robots. Users can utilize RAPID to set up and execute their workflows and

> processes. This is enabled by user-defined libraries called **Modules**, that contain variables and functions or
> processes **(PROCs)**. Modules can then be loaded in controller **Tasks**, and called as required.

Reviewer: ```suggestion
processes. This is enabled by user-defined libraries called *Modules*, that contain variables and functions or
processes (*PROCs*). Modules can then be loaded in controller *Tasks*, and called as required.
```

experiment. After correct operation is verified, users can just switch to the real hardware interface to connect to the
actual robot or simulator.

:::note

The mock interface does not include a physics engine, or dynamics calculations, and merely sets the commands as state.
For more sophisticated simulation capabilities, check out RobotStudio in the following section.

:::

## Connecting to a robot

The provided hardware interface can be used to connect and control either a simulated or a real robot. The following
sections provide the necessary steps to do both.

### RobotStudio simulation

[RobotStudio](https://new.abb.com/products/robotics/software-and-digital/robotstudio) is the official ABB offline
programming and simulation tool for robotics applications. It allows to run virtual controllers that mimic the behavior
of the real robot, ensuring seamless transition between simulation and hardware. It can also be used to configure
several aspects of the real controller.

:::note

The RobotStudio software suite is available only for Windows, so you may need to set it up on a secondary device or a
virtual machine .

:::

> Setting up a virtual workstation and controller is the next step after using the mock interface. This can be achieved by

Reviewer: Here too, until here no mention of interfaces so i'd prefer to rephrase

following the next steps:

1. In RobotStudio, navigate to the **Add-Ins** tab and go to **Gallery**. There is a list of all available robot models

>    and RobotWare versions, the internal controller software. Make sure to install the versions present in the actual

Reviewer: maybe replace "actual" with something saying "To ensure consistency between simulation and reality, install the versions matching the real robot controller, if one is available.

   robot controller, to ensure consistency between simulation and reality.

> <div class="text--center">
>   <img src={abbInstallAddins} alt="Install necessary addins in RobotStudio." />
> </div>

Reviewer: ```suggestion
    <div class="text--center">
      <img src={abbInstallAddins} alt="Install necessary addins in RobotStudio." />
    </div>
```
This aligns the left border of the picture with the text, same below

2. Go back to to **File > New > Project**.

> 3. Select to create a new controller and define the robot model and variant, as well as the RobotWare version.

Reviewer: I actually didn't realize this, should the RobotWare 7.X be our minimal requirement?

Author: Finding a single line of useful information in ABB documentation is a nightmare. Including it here for future reference:

https://developercenter.robotstudio.com/webservice

Here we read:

_...Robot Web Services 2.0 is available in RobotWare 7 which ships with our new OmniCore controller generation. For older RobotWare versions you can use Robot Web Services 1.0. It is deprecated in RobotWare 7._

To sum up, yes probably we should have RobotWare 7.X as a minimal requirement. I will add it somewhere.

Author: Note on version added.

4. Make sure to activate the **Customize Options** button. This is required to add EGM in a next step.

<div class="text--center">
  <img src={abbNewProject} alt="Create a new RobotStudio project." />
</div>

5. Select **Create** to create the new project.

> 6. In the window that pops up, in the **Options** tab, look for **EGM** and **RobotStudio Connect** and add them in the

Reviewer: RobotStudio Connect seems to be a requirement but is not mentioned above under prerequisites. This can be relevant because it might have to be purchased with the robot. Do you know what it is for?

Author: Really struggled to find information on this:

<img width="889" height="867" alt="image" src="https://github.com/user-attachments/assets/0546daa3-d769-4fcc-8984-2e768f81164d" />

and forum post where it is mentioned that it is a necessary feature:

https://forums.robotstudio.com/discussion/13321/pcsdk-or-rws

The reason I am not expanding too much on RobotStudio Connect is that I am not really sure what to say, other that it is necessary to connect to the robot, which is quite generic. Meaning that I am not sure what it would affect if it was missing: full connection with EGM, RWS, access to the controller? Not sure.

   controller. Then select **Apply and Reset** to finalize.

<div class="text--center">
  <img src={abbAdditionalOptions} alt="Additional options in the RobotStudio project." />
</div>

> ### Omnicore controller & real robot

Reviewer: I'm not sure how "Omnicore" is relevant here, if we clarify before.

:::

### RWS

The second ABB feature that AICA system utilizes to connect to the robot is RobotWebServices (RWS). RWS is a platform
that enables developers to create applications that interact with the robot controller, using RESTful APIs that leverage
the HTTPS protocol. The hardware interface uses RWS for auxiliary functionality, such as starting/stopping the program
and the motors, and setting IOs. Setting up RWS in the simulator and the actual robot requires slightly different steps,
which will be explained in the following sections. More information can be found in the
[product reference page](https://developercenter.robotstudio.com/api/RWS).

### RAPID

RAPID is the programming language of ABB robots. Users can utilize RAPID to set up and execute their workflows and
processes. This is enabled by user-defined libraries called **Modules**, that contain variables and functions or
processes **(PROCs)**. Modules can then be loaded in controller **Tasks**, and called as required.

<!-- :::note

The example above drives the robot in velocity by setting the **PosCorrGain** parameter to 0 in the control loop. Remove
this if the robot needs to receive position commands.

::: -->

> ## Mock interface

Author: To be accurate, becuase I just saw it, mock interface is mentioned in the [An application with hardware](https://docs.aica.tech/docs/examples/guides/mock-hardware-example) guide. If someone thinks it's cluttering the page it can be removed, but I would leave it in for completeness.

> ## Mock interface

Reviewer: I mentioned this before, I still think that is out of place here. We're doing very ABB specific things before and after so if anything, it should come as a tip below maybe before starting an actual application

restart the controller.

<div class="text--center">
  <img src={abbControllerConfiguration} alt="Controller configuration settings." />
</div>

<div class="text--center">
  <img src={abbNewUDPUCDevice} alt="Add a new UDPUS device." />
</div>

:::tip

While this takes seconds in simulation, the restart procedure in the real robot might take a few minutes, so make all necessary 
changes, after the next section, and then restart.

:::

Next step is enabling RWS connection. For the simulation specifically, this requires either using a proxy or whitelisting the IP address of the device
trying to access RWS, in this case the device running the AICA application. The first approach is preferable and
described analytically in a
[RobotStudio forum post](https://forums.robotstudio.com/discussion/12082/using-robotwebservices-to-access-a-remote-virtual-controller)
(read until the end and the last comment for a critical fix). Omnicore controllers and RobotWare 7.x versions by default
listen on HTTPS and port (80 for RobotStudio and 443 for the real robot). This can be modified by following the instructions in this

> [forum post](https://forums.robotstudio.com/discussion/12177/how-to-change-the-listening-port-of-the-virtual-controller-robotware-6-x-and-7-x). 

Author: Writing here, but it is a comment for both. In general I would be fully in favor of duplicating the instructions, since we remove external dependency. However, there are several steps and it might get extensive and cluttered. If I started removing some of them, I would still have to reference the original posts for completeness, or to give credit as a matter of fact. So I decided to leave it like that.

> [forum post](https://forums.robotstudio.com/discussion/12177/how-to-change-the-listening-port-of-the-virtual-controller-robotware-6-x-and-7-x). 

Reviewer 2: I see, yeah that's understandable. If that's the case and the others prefer to not duplicate, then the rest of the wording is good for me

                  \MaxSpeedDeviation:=120.0;

      WHILE TRUE DO
          ! Start the EGM communication session.
          EGMRunJoint egm_id, EGM_STOP_HOLD, \J1 \J2 \J3 \J4 \J5 \J6 \CondTime:=5 \RampOutTime:=5 \PosCorrGain:=0;
          EGMStop egm_id, EGM_STOP_HOLD \RampOutTime:=5;
      ENDWHILE
      ! Release the EGM id.
      EGMReset egm_id;

  ERROR
      IF ERRNO = ERR_UDPUC_COMM THEN
          TPWrite "Communication timedout";
          TRYNEXT;
      ENDIF
  ENDPROC
ENDMODULE
```

</details>

## Hardware interface

Returning to AICA studio and the hardware interface, it is now possible to define the parameters and connect to the
robot. The majority of the hardware interface parameters enable connection to EGM and RWS:

- EGM port: the port that EGM uses to send commands.
- RWS port & IP: port and address of the RWS.
- Connection timeout: for RWS connection.
- Rapid File Path: the path for the module to be loaded in the controller home directory.
- Headless Mode: if true, the hardware interface handles the whole initialization procedure of starting and stopping the
RAPID program and the motors without requiringn any user interaction.

> The EGM port should be set to the same value defined in the UDPUC device in the controller

Author: Hmm, maybe it would be good idea, providing here the actual values that I used through the guide (since this is the very last part, there is nothing to add after).  This would also explain the 9876 port that appears in the next screenshot for RWS that is not explained elsewhere.

   the external control device, so the address should be set accordingly.

<div class="text--center">
  <img src={abbControllerConfiguration} alt="Controller configuration settings." />
</div>

<div class="text--center">
  <img src={abbNewUDPUCDevice} alt="Add a new UDPUS device." />
</div>

2. Next step is enabling RWS connection. For the simulation specifically, this requires either using a proxy or
   whitelisting the IP address of the device trying to access RWS, in this case the device running the AICA application.
   The first approach is preferable and described analytically in a
   [RobotStudio forum post](https://forums.robotstudio.com/discussion/12082/using-robotwebservices-to-access-a-remote-virtual-controller)
   (read until the end and the last comment for a critical fix). Omnicore controllers and RobotWare 7.x versions by
   default listen on HTTPS and port 80 for RobotStudio and 443 for the real robot. This would not really be necessary,
   but if needed, it can be modified by following the instructions in this
   [forum post](https://forums.robotstudio.com/discussion/12177/how-to-change-the-listening-port-of-the-virtual-controller-robotware-6-x-and-7-x)

>    (set to 9876 in the example at the end of this guide).

Reviewer: Yeah, maybe something like that
<img width="1916" height="955" alt="image" src="https://github.com/user-attachments/assets/9fe06ddc-ba20-410b-825e-ed755c06a47e" />

>    (set to 9876 in the example at the end of this guide).

Author: I did not like how the screenshot looks there, because we are only showing RobotStudio at that point. I have added it in the last section of the Hardware Interface. Let me know what you think.
